torch/csrc/jit/frontend/ir_emitter.cpp

```diff
@@ -3303,6 +3341,14 @@ struct to_ir {
       const TypePtr& type_hint = nullptr) {
     switch (tree.kind()) {
       case TK_VAR: {
+        // There are certain cases in which we know a variable is of
+        // type `Union[T1, T2, ..., TN]`, but the type of the variable's
+        // value is still `T1` (one of the types in the Union). This
+        // correctly sets the value's type to match the variable's
+        // Union annotation
+        if (type_hint && type_hint->kind() == UnionType::Kind) {
+          environment_stack->setType(Var(tree).name().name(), type_hint);
```

Comment:
this api is bypassing the safety checks. I think we should be using setSugaredVar or something along those lines. I'm also a little confused about what we're trying to achieve here. Do you mind commenting on the use case ?